puppetry/__init__.py

```diff
@@ -238,7 +251,46 @@ def enable_parts(args):
         parts_mask = int(new_parts_mask)
         _logger.info(f"enable_parts set mask to {parts_mask}")
 
-@registerCommand("set_skeleton")
+@registerCommand('joint_report')
+def report(args):
+    '''Receive a report of post-ik data for a joint'''
+
+    global _report_data
+
+    #Add the report for this joint to the _report_data.
+    if args is not None:
+        joint_id = args['joint_id']
+        _report_data[joint_id] = args
+
+
+def flush_reports( joints = None ):
+    '''Flushes report_data.
+        If joints is None, removes all.
+        Otherwise, joints is assumed to be an array of joints to flush.
+    '''
+
+    global _report_data
+
+    if joints is None:
+        _report_data = {}
+    else:
+        for joint in joints:
+            _report_data.pop(int(joint))
+
+def get_report( joint_id = None):
+    '''Accessor for _report_data
+       Returns the report for the joint ID if it is an integer or None if not found
+       returns entire _report_data structure if joint_id is None'''
+
+    if joint_id is not None:
+        jid = int(joint_id)
+        if jid in _report_data:
```

Comment:
Exceptions in Python don't incur the disproportionate overhead of exceptions in C++. In the general case, it's been recommended to write something more like this:

    try:
        return _report_data[jid]
    except KeyError:
        return None

because attempting to avoid an exception with a prior `jid in _report_data` test requires two distinct `dict` lookups, whereas the code above requires only one.

But for a default value from a `dict` lookup, you could just call `dict.get()`:

    return _report_data.get(jid, None)

or even simpler, since `None` is the default default:

    return _report_data.get(jid)

Reply:
Thank you for taking the time to write that out. I'll adopt these strategies going forward.